[FLINK-11403][network] Introduce ResultPartitionWithConsumableNotifier in task for notifying consumable result partition

zhijiangW opened the pull request:
## What is the purpose of the change

*The consumable notification shuould belong to the scheduler scope, so it is not general for all the `ResultPartitionWriter` implementations. The current creation of `ResultPartitionWriter` from `NetworkEnvironment` relies on `JobID`, `TaskAction`, `ResultPartitionConsumableNotifier` for handling the consumable notification logic. For breaking this tie to make `ShuffleService#createResultPartitionWriters` simple, `Task` could wrap `ResultPartitionWriter`
into `ResultPartitionWithConsumableNotifier` for handling the notification.*

## Brief change log

  - *Wrap `ResultPartitionWithConsumableNotifier` in task for handing the notification logic*
  - *Remove the notification logic from `ResultPartition`*

## Verifying this change

This change is already covered by existing tests, such as `ResultPartitionTest` .

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented:
FYI, next I would remove `owningTaskName`, `JobID` and `TaskActions` from the constructor of `ResultPartition`. And one possible way is introducing `registerTaskActions(String owningTaskName,TaskActions taskActions)` outside of constructor like current `registerBufferPool`, then it makes clean for the constructor to be created from `ShuffleService` in future.

zhijiangW commented:
FYI, this PR is based on #7593 which could get the flag of `sendScheduleOrUpdateConsumers` from TDD directly, so I include that commit.
I would add some new tests in `RecordWriterTest` to verify the logic for `notifyPipelinedConsumers` after confirming the main logic changes make sense.

zhijiangW commented:
@azagrebin , thanks for your reviews!

I think your suggestion makes sense and is worth trying. We wrap the `ResultPartition` created from `ShuffleService` and other fields related to consumption notification in a separate new class. Then the `RecordWriter` and `Task` only see this new wrapped class avoid of spreading tiny ones anywhere.

I would have a try and re-submit the codes when ready.

zhijiangW commented:
@azagrebin, I updated the codes based on our latest comments.

zhijiangW commented:
Also, I would add the new `ResultPartitionWriterTest` for verifying the logic after confirming this way right.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriterWrapper.java:
```diff
@@ -0,0 +1,100 @@
+ * A wrapper of result partition writer for handling notification of the consumable
+ * partition which is added a {@link BufferConsumer} or finished.
+ */
+public class ResultPartitionWriterWrapper {
```
Do we need to expose this class to all touched classes? If it would implement `ResultPartitionWriter` interface, we could just wrap it once in `Task` where it is currently created. `ResultPartitionWriter` could be used everywhere because other classes are mostly interested in `ResultPartitionWriter` but not in notification specifics.

zhijiangW replied:
Actually only `RecordWriter` and `Task` need touch this wrapper class for the requirement of notification in `addBufferConsumer` and `finish`. 

- Task calls the `finish` logic, so it has to maintain `ResultPartitionWriterWrapper` for this notification.

- `ResultPartitionWriterWrapper` has to be passed into `Environment` in order to be got for creating `RecordWriter`.

`Task` temporarily also maintains the previous array of `ResultPartition` which would be changed after refactoring the current `ResultPartitionWriter` interface. Do I get your points or misunderstanding?

zhijiangW commented:
@azagrebin , I renamed the classes of `NetworkResultPartition` and `ResultPartition` as we confirmed , and make `ResultPartition` also implement the `ResultPartitionWriter`.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -19,432 +19,105 @@
-	private final ResultSubpartition[] subpartitions;
-
-	private final ResultPartitionManager partitionManager;
+	private final ResultPartitionWriter partitionWriter;
```
At the moment, up to this PR, `ResultPartitionWriter` interface is not abstracted fully yet.
So we see this PR as just one step closer to the refactoring of `ResultPartitionWriter` interface.

I would keep this field explicitly as `NetworkResultPartition` and keep a getter method for it.
Then we can keep `ResultPartition` where possible outside of `NetworkEnviroment` and whenever we need to call a specific method of `NetworkResultPartition` (not `ResultPartitionWriter`), 
we can use `resultPartition.getNetworkResultPartition().specificMethod()`.

Later, we can quickly drop `getNetworkResultPartition()` method and see want we miss to switch from `ResultPartition` outside of `NetworkEnviroment` to `ResultPartitionWriter`.

This way we could have less changes in this PR. Especially, we do not need a `NetworkResultPartition` array in Task.

What do you think?

zhijiangW replied:
In the final form, `ResultPartition` should not implement `ResultPartitionWriter` and it could be put in the `TaskManager` package instead of current network package. And the field in `ResultPartition` should be regular `ResultPartitionWriter` finally.

I could understand your concern. If we keep special `NetworkResultPartition` in `ResultPartition` temporarily, it would bring other troubles after I have a try. The mainly involved ones are `Environment` and `RecordWriter`.

- `RecordWriter` -> `ResultPartition` -> `NetworkResultPartition`. There are many other implementations for `ResultPartitionWriter` used for tests. In `RecordWriterTest` we could not construct the proper `RecordWriter` based on `ResultPartition` which contains specific `NetworkResultPartition`.

- `Environment` should keep `ResultPartition` instead of `ResultPartitionWriter`, because it would be used for constructing `RecordWriter` later. If we keep specific `NetworkResultPartition` in `ResultPartition`, then it could not return proper `ResultPartition` for other `Environment` implementations, such as `StreamMockEnvironment`, `MockEnvironment` etc, because they maintain other `ResultPartitionWriter` implementations.

So I think the current way is more easy to handle, because we only need refactor `ResultPartition` and `Task` finally. Another option is we suspend this PR currently, you could firstly review other PRs which are involved in `ResultPartitionWriter` interface refactor. After we make a final interface form, then we could come back to maintain `ResultPartitionWriter` in `ResultPartition` finally. What do you think?

zhijiangW replied:
I checked the other related PRs. The final form of `ResultPartitionWriter` interface would only cover some basic methods of `getBufferProvider`, `addBufferConsumer`, `flushAll`, `flush` after merging #7822 to rebase. Then the constructor of `ResultPartition` would cover some other fields such as `numberOfSubpartitions`, `numberOfTargetKeyGroups` which could be got directly from `ResultPartitionDeploymentDescriptor`.

azagrebin replied:
Why should `ResultPartition` (notification) not implement eventually `ResultPartitionWriter`?
Then everything could depend just on `ResultPartitionWriter` interface.
The `ResultPartition` could become:
`ResultPartitionWriterWithConsumableNotification implements ResultPartitionWriter` in Task package.
Task could just wrap `ShuffleService.ResultPartitionWriter` (`NetworkResultPartition` for netty) with `ResultPartitionWriterWithConsumableNotification` and pass it to `RecordWriter` also as `ResultPartitionWriter`.
Also tests could do the same if notification is relevant for them or use just `NetworkResultPartition` as `ResultPartitionWriter`.

but I think I see your concern at the moment, as we do not keep `NetworkResultPartition` per Task in `NetworkEnviroment` yet, probably having 2 arrays for writers is the simplest at this step.
We can also work on other PRs and see how it goes, maybe it will simplify this one.

zhijiangW replied:
It is no problem to let `ResultPartition` also implement `ResultPartitionWriter`. And it might also make sense to not implement the interface. Because `ResultPartition` would be refactored to task package finally, if it implements the interface, it seems still coupled with network interface. In addition, there already exists a `ResultPartitionWriter` component in `ResultPartition`. If the `ResultPartition` itself is also a `ResultPartitionWriter`, it might seem a bit duplicated.

ShuffleService would create regular `ResultPartitionWriter` which is wrapped  into `ResultPartitionWithConsumableNotification` in task stack. And ` ResultPartitionWithConsumableNotification` would be passed into `RuntimeEnvironment` to be used for creating `RecordWriter` later. It might also makes sense to reference with `ResultPartitionWithConsumableNotification` instead of `ResultPartitionWriter` in `RecordWriter` and `RuntimeEnvironment`. 

As you said, we could keep the current status. And after other parts ready, we could further check how it goes. :)

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/NetworkResultPartition.java:
```diff
@@ -0,0 +1,414 @@
+		this.bufferPool = checkNotNull(bufferPool);
+	}
+
+	public String getOwningTaskName() {
```
can be package-private

zhijiangW replied:
Yes, I just refactored the class name, and this is the previous existing issue. I could fix it in this PR together.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/NetworkResultPartition.java:
```diff
@@ -0,0 +1,414 @@
+
+	private final ResultPartitionManager partitionManager;
+
+	public final int numTargetKeyGroups;
```
can be private

zhijiangW replied:
ditto

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/NetworkResultPartition.java:
```diff
@@ -0,0 +1,414 @@
+	 * <p>The partition can only be released after each subpartition has been consumed once per pin
+	 * operation.
+	 */
+	void pin() {
```
can be private

zhijiangW replied:
ditto

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -19,432 +19,105 @@
-
-	// ------------------------------------------------------------------------
+	public void finish() throws IOException {
+		partitionWriter.finish();
```
it seems we also need to check success of `partitionWriter.finish()` to preserve the old behaviour.
Maybe, `ResultPartitionWriter.finish()` could also return boolean.

zhijiangW replied:
Yes, it is better to keep consistent with previous behavior. I would fix it.

azagrebin replied:
True, that previous try/finally is not needed actually, it confused me :)

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -19,432 +19,105 @@
-	private static final Logger LOG = LoggerFactory.getLogger(ResultPartition.class);
-
-	private final String owningTaskName;
+public class ResultPartition implements ResultPartitionWriter {
```
After switching to `ResultPartitionWriter` fully, this should be renamed to something like `ResultPartitionWithConsumableNotification` then it probably does not belong to network specific package anymore but maybe rather to `Task` package.

zhijiangW replied:
Yes, I agree with your point. 
But currently it still implements `ResultPartitionWriter`, so we could keep it as current way? In future this class might not need implement the interface, then we could further refactor it as you said. Or you think we should rename this class and change the package now?

azagrebin replied:
I think we can keep it for now as it is in this PR, because `ResultPartitionWriter` is not fully abstracted yet, just thought for `ResultPartitionWriter` refactoring PR.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionTest.java:
```diff
@@ -148,15 +148,17 @@ public void testAddOnReleasedBlockingPartition() throws Exception {
+	 * @param partitionType the result partition type to set up
 	 */
-	protected void testAddOnReleasedPartition(final ResultPartitionType pipelined)
+	protected void testAddOnReleasedPartition(final ResultPartitionType partitionType)
```
can be private, also some other methods in this class

zhijiangW replied:
ditto

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionTest.java:
```diff
@@ -67,7 +67,7 @@ public void testSendScheduleOrUpdateConsumersMessage() throws Exception {
 			verify(notifier, times(1))
 				.notifyPartitionConsumable(
-					eq(partition.getJobId()),
+					any(JobID.class),
```
I guess we could still keep `@VisibleForTesting ResultPartition.getJobId()`?

zhijiangW replied:
Yes, it can be package private method.

azagrebin commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionTest.java:
```diff
@@ -148,15 +148,17 @@ public void testAddOnReleasedBlockingPartition() throws Exception {
-			ResultPartition partition = createPartition(notifier, pipelined, true);
-			partition.release();
+			NetworkResultPartition networkPartition = createNetworkResultPartition(partitionType);
+			ResultPartition partition = new ResultPartition(
```
do we need `ResultPartition` in this test?
`networkPartition. addBufferConsumer` seems to work as well.

zhijiangW replied:
I think we might still need the `ResultPartition`.
This test wants to verify the notification logic after releasing partition. The `partition.addBufferConsumer` would return false after released, then it would never trigger notification which is verified in finally path. And the notification logic is removed from `networkPartition`, so it can not be verified directly.

azagrebin replied:
agree

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -236,21 +237,23 @@ protected ResultPartitionWriter createResultPartition(
+			ioManager);
+		environment.setupPartition(networkResultPartition);
+
+		ResultPartition resultPartition = new ResultPartition(
```
if this `resultPartition` has `NoOpTaskActions` and `NoOpResultPartitionConsumableNotifier`, do we still need to wrap `networkResultPartition` with it? Could we just return `networkResultPartition`?

zhijiangW replied:
From functional aspect, I agree with your point.  But it seems reasonable to reflect the real path for performance sensitive benchmark. If there are extra overheads executed in `ResultPartition`, it could be detected in the benchmark. What do you think?

azagrebin replied:
ok, let's keep it

zhijiangW commented:
@azagrebin , thanks for reviews! I submitted a separate fixup commit for addressing the method access modifier. For the other issues I left some comments need to be further confirmed.

zhijiangW commented:
@azagrebin I have rebased the latest codes and addressed the previous comments.

FYI, the previous arrays of `ResultPartition` in task is only temporarily used for `NetworkEnvironment#registerTask` now, and it would be removed in #8310 , then the task only needs maintain the arrays of `ResultPartitionWriter`.

zhijiangW commented:
@azagrebin I have rebased the master to solve the conflicts.

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -370,16 +370,26 @@ public Task(
 
+		int counter = 0;
+		this.producedPartitions = new ResultPartitionWriter[resultPartitionDeploymentDescriptors.size()];
+		for (ResultPartitionDeploymentDescriptor desc : resultPartitionDeploymentDescriptors) {
```
nit: suggest to move wrapping loop into a separate static method

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -272,8 +239,6 @@ public void finish() throws IOException {
 		finally {
 			if (success) {
```
side comment: strange `success` flag looks redundant, also try/catch block ...

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java:
```diff
@@ -45,7 +45,7 @@
 
 	/**
-	 * Adds the bufferConsumer to the subpartition with the given index.
+	 * Adds the bufferConsumer to the subpartition with the given index and returns whether successful or not.
```
`whether it was added successfully or not`

azagrebin commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java:
```diff
@@ -0,0 +1,145 @@
+
+	private final ResultPartitionConsumableNotifier partitionConsumableNotifier;
+
+	private final boolean sendScheduleOrUpdateConsumersMessage;
```
I would suggest we do not have this flag `sendScheduleOrUpdateConsumersMessage` here but use it as a condition whether to wrap original `ResultPartitionWriter` in the `Task` constructor at all because it seems we do nothing here if `sendScheduleOrUpdateConsumersMessage` is `false`.

zhijiangW replied:
very good suggestion.

azagrebin commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -213,7 +213,6 @@ protected ResultPartitionWriter createResultPartition(
 			int channels) throws Exception {
 
 		ResultPartition resultPartition = new ResultPartitionBuilder()
-			.setJobId(jobId)
```
We should actually also wrap `resultPartition` here to take wrapping into account in benchmark tests. Probably with `NoOpTaskActions`, at least forcing the wrapper to be called.

zhijiangW commented:
Thanks for the further review @azagrebin !

I pushed a fixup commit for addressing above comments.

zhijiangW commented:
Thanks for helping trigger the benchmark @azagrebin .
I pushed a fixup commit for addressing the new comments.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -258,24 +225,13 @@ public void flush(int subpartitionIndex) {
+		for (ResultSubpartition subpartition : subpartitions) {
+			subpartition.finish();
 		}
-		finally {
```
Could you double check if there are no `RuntimeException`s that we are expecting to be thrown from `subpartition.finish()`? I think that would be the only reason for having this strange `try/finally/success` thingy.

zhijiangW replied:
Yes, I also checked the related processes and it seems no explicit `RuntimeException` here.

zhijiangW commented:
Thanks for reviews @pnowojski !

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java:
```diff
@@ -0,0 +1,174 @@
+ * results, receivers are deployed as soon as the first buffer is added to the result partition.
+ * With blocking results on the other hand, receivers are deployed after the partition is finished.
+ */
+public class ResultPartitionWithConsumableNotifier implements ResultPartitionWriter {
```
Shouldn't this class better be called `ConsumableNotifyingResultPartitionWriterDecorator` or something else?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java:
```diff
@@ -0,0 +1,174 @@
+
+	public static ResultPartitionWriter[] create(
+			Collection<ResultPartitionDeploymentDescriptor> descs,
+			ResultPartitionWriter[] partitions,
```
better to call `partitionWriters`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -190,7 +190,7 @@
 	private final SerializedValue<ExecutionConfig> serializedExecutionConfig;
 
-	private final ResultPartitionWriter[] producedPartitions;
+	private final ResultPartitionWriter[] partitionWithNotifiers;
```
I think the field name could rather be `resultPartitionWriters` or `consumableNotifyingPartitionWriters`.

zhijiangW replied:
I would take `consumableNotifyingPartitionWriters`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -369,16 +369,20 @@ public Task(
 
 		// produced intermediate result partitions
-		this.producedPartitions = networkEnvironment.createResultPartitionWriters(
+		final ResultPartitionWriter[] partitions = networkEnvironment.createResultPartitionWriters(
```
`resultPartitionWriters`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/api/writer/AbstractCollectingResultPartitionWriter.java:
```diff
@@ -70,10 +70,11 @@ public BufferBuilder getBufferBuilder() throws IOException, InterruptedException
+	public synchronized boolean addBufferConsumer(BufferConsumer bufferConsumer, int targetChannel) throws IOException {
 		checkState(targetChannel < getNumberOfSubpartitions());
-		bufferConsumers.add(bufferConsumer);
+		boolean success = bufferConsumers.add(bufferConsumer);
```
why do we return this value as the method's return value?

zhijiangW replied:
I think it has two possibilities here.  One is return true, and the other is throwing an exception. I could return true directly. WDYT?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java:
```diff
@@ -0,0 +1,174 @@
+	//  Factory
+	// ------------------------------------------------------------------------
+
+	public static ResultPartitionWriter[] create(
```
could be named `decorate` instead of `create`

tillrohrmann commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java:
```diff
@@ -213,8 +215,7 @@ protected ResultPartitionWriter createResultPartition(
 
-		ResultPartition resultPartition = new ResultPartitionBuilder()
-			.setJobId(jobId)
+		ResultPartitionWriter partition = new ResultPartitionBuilder()
```
name should be `resultPartitionWriter`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/ResultPartitionWithConsumableNotifier.java:
```diff
@@ -0,0 +1,174 @@
+	 * this will trigger the deployment of consuming tasks after the first buffer has been added.
+	 */
+	private void notifyPipelinedConsumers() {
+		if (!hasNotifiedPipelinedConsumers && partitionType.isPipelined()) {
```
can't we check `partitionType.isPipelined` when we create an instance of this class? That way we would not have to make this case distinction here.

zhijiangW commented:
@tillrohrmann thanks for your reviews and I have pushed one fixup commit for addressing above comments!

zhijiangW commented:
Thanks for confirmation @tillrohrmann .
The travis fails because of table issues which seem not relevant with my changes because I saw other PRs also fail with the same table issues. Do you think I should re-run the travis to make it green?

tillrohrmann commented:
Quick question @zhijiangW, did you change anything in the last force push? You overwrote my force push where I had everything squashed and now I'm unsure whether my local branch reflects the same state as your latest push. I would suggest to not force push after I say that I merge the PR. I usually take the commits, squash them, adjust the commit message and then force push back into the contributor's feature branch to trigger another Travis run. If Travis is green, then I merge this state.

tillrohrmann commented:
Or was it simply to retrigger Travis?

zhijiangW commented:
I just solved the conflicts because of my another PR is merged by piotr, not changed anything for this PR. Sorry for not confirming with you before, and I am now clear of the process.

zhijiangW commented:
I just solved the conflicts because of my another PR is merged by piotr, not changed anything for this PR. Sorry for not confirming with you before, and I am now clear of the process.

tillrohrmann commented:
Makes sense. Thanks a lot @zhijiangW. Travis gave green light. Merging this PR now :-)
